src/chunks.rs

```diff
@@ -0,0 +1,191 @@
+use anyhow::Context;
+use std::fs::File;
+use std::io::{BufRead, BufReader, Cursor, Seek, SeekFrom};
+use std::path::Path;
+
+fn split(chunk: u64, size: u64) -> Vec<u64> {
+    let e = size / chunk + 1.min(size % chunk);
+    (0..e).map(|i| i * chunk).collect()
+}
+
+#[derive(Debug)]
+pub struct Chunk<C> {
+    lines: std::io::Lines<C>,
+    pos: u64,
+    end: u64,
+}
+
+impl<C> Chunk<C> {
+    pub fn new(
+        mut c: C,
+        chunk: usize,
+        mut current: u64,
+        start: u64,
+        size: u64,
+    ) -> anyhow::Result<(Self, u64)>
+    where
+        C: Seek + BufRead,
+    {
+        let skip = if current > start {
+            true
+        } else if start != current {
+            c.seek(SeekFrom::Start(start - 1))?;
+            let mut buf = [0 as u8; 1];
+            if let Ok(1) = c.read(&mut buf) {
+                buf[0] != b'\n'
+            } else {
+                false
+            }
+        } else {
+            false
+        };
+
+        c.seek(SeekFrom::Start(start))?;
+        current = if skip {
+            let mut skip_leader = String::new();
+            let _ = c.read_line(&mut skip_leader)?;
+            start + skip_leader.len() as u64
+        } else {
+            start
+        };
+        let lines = c.lines();
+        let c = Self {
+            lines,
+            pos: current,
+            end: size.min(start + chunk as u64),
+        };
+        Ok((c, current))
+    }
+}
+
+impl<T> Iterator for Chunk<T>
+where
+    T: Seek + BufRead,
+{
+    type Item = String;
+    fn next(&mut self) -> Option<String> {
+        if self.pos >= self.end {
+            return None;
+        }
+        match self.lines.next() {
```

Comment:
You can use the try operator `?` to return `None` instead of matching explicitly (so this should translate to `let l = self.lines.next()?.ok()?;`.

Reply:
Done. That's much better.